src/main/java/com/limechain/grandpa/GrandpaService.java

```diff
@@ -31,6 +31,93 @@ public GrandpaService(GrandpaSetState grandpaSetState, BlockState blockState) {
         this.blockState = blockState;
     }
 
+    /**
+     * Determines if the specified round can be finalized.
+     * 1) Checks for a valid preVote candidate and ensures it's completable.
+     * 2) Retrieves the best final candidate for the current round, archives it,
+     * and compares it to the previous round’s candidate.
+     *
+     * @param roundNumber
+     * @return if given round is finalizable
+     */
+    private boolean isFinalizable(GrandpaRound grandpaRound) {
+
+        Vote preVoteCandidate = getGrandpaGhost(grandpaRound);
+        if (preVoteCandidate == null) {
+            return false;
+        }
+
+        grandpaRound.setPreVotedBlock(preVoteCandidate);
+
+        if (!isCompletable(grandpaRound)) {
+            return false;
+        }
+
+        Vote bestFinalCandidate = getBestFinalCandidate(grandpaRound);
+        if (bestFinalCandidate == null) {
+            return false;
+        }
+
+        grandpaRound.setBestFinalCandidate(bestFinalCandidate);
+
+        var prevGrandpaRound = grandpaRound.getPrevious();
+        if (prevGrandpaRound == null) {
+            return false;
+        }
+
+        Vote prevBestFinalCandidate = prevGrandpaRound.getBestFinalCandidate();
+
+        return prevBestFinalCandidate != null
+                && prevBestFinalCandidate.getBlockNumber().compareTo(bestFinalCandidate.getBlockNumber()) <= 0
+                && bestFinalCandidate.getBlockNumber().compareTo(preVoteCandidate.getBlockNumber()) <= 0;
+    }
+
+    /**
+     * To decide if a round is completable, we need two calculations
+     * 1. [TotalPcVotes + TotalPcEquivocations > 2/3 * totalValidators]
+     * 2. [TotalPcVotes - TotalPcEquivocations - (Votes where B` > Ghost) > 2/3 * totalValidators]
+     * Second calculation should be done for all Ghost descendants
+     *
+     * @param grandpaRound
+     * @return if the current round is completable
+     */
+    private boolean isCompletable(GrandpaRound grandpaRound) {
+
+        Map<Vote, Long> votes = getDirectVotes(grandpaRound, Subround.PRECOMMIT);
+        long votesCount = votes.values().stream()
+                .mapToLong(Long::longValue)
+                .sum();
+
+        long equivocationsCount = grandpaRound.getPcEquivocationsCount();
+        long totalVoters = grandpaSetState.getAuthorities().size();
+        long threshold = (2 * totalVoters) / 3;
+
+        if (votesCount + equivocationsCount <= threshold) {
+            return false;
+        }
+
+        List<Vote> ghostDescendents = getBlockDescendents(
+                grandpaRound.getPreVotedBlock(),
+                new ArrayList<>(votes.keySet())
+        );
+
+        for (Vote vote : ghostDescendents) {
+
+            var currentBlockHash = vote.getBlockHash();
+            var observedVotesForCurrentBlock = getObservedVotesForBlock(
```

Comment:
```suggestion
            var descendantBlockHash = vote.getBlockHash();
            var observedVotesForDescendantBlock = getObservedVotesForBlock(
```
I think this is more explicit regarding which block we are referring to.